// posthog-js/lib/src/entrypoints/logs.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
var api_logs_1 = require("@opentelemetry/api-logs");
var exporter_logs_otlp_http_1 = require("@opentelemetry/exporter-logs-otlp-http");
var sdk_logs_1 = require("@opentelemetry/sdk-logs");
var resources_1 = require("@opentelemetry/resources");
var globals_1 = require("../utils/globals");
var core_1 = require("@posthog/core");
var setupOpenTelemetry = function (posthog) {
    var _a;
    var serviceName = ((_a = posthog.config.logs) === null || _a === void 0 ? void 0 : _a.serviceName) || 'posthog-browser-logs';
    var attributes = {
        'service.name': serviceName,
        host: globals_1.assignableWindow.location.host,
    };
    if (posthog.sessionManager) {
        var _b = posthog.sessionManager.checkAndGetSessionAndWindowId(true), sessionId = _b.sessionId, windowId = _b.windowId;
        attributes = __assign(__assign({}, attributes), { 'session.id': sessionId, 'window.id': windowId });
    }
    api_logs_1.logs.setGlobalLoggerProvider(new sdk_logs_1.LoggerProvider({
        resource: (0, resources_1.resourceFromAttributes)(attributes),
        processors: [
            new sdk_logs_1.BatchLogRecordProcessor(new exporter_logs_otlp_http_1.OTLPLogExporter({
                url: "".concat(posthog.config.api_host, "/i/v1/logs?token=").concat(posthog.config.token),
                // 1. Force the content type to text/plain to avoid OPTIONS preflight
                headers: {
                    'Content-Type': 'text/plain',
                },
            })),
        ],
    }));
};
var LOG_BODY_SIZE_LIMIT = 10000;
var LOG_ATTRIBUTES_LIMIT = 50;
var appendWithLimit = function (parts, text, budget) {
    if (budget.remaining <= 0) {
        budget.truncated = true;
        return false;
    }
    if (text.length <= budget.remaining) {
        parts.push(text);
        budget.remaining -= text.length;
        return true;
    }
    parts.push(text.slice(0, budget.remaining));
    budget.remaining = 0;
    budget.truncated = true;
    return false;
};
var stringifyStringWithLimit = function (value, parts, budget) {
    var serialized = JSON.stringify(value);
    if (serialized.length <= budget.remaining) {
        return appendWithLimit(parts, serialized, budget);
    }
    budget.truncated = true;
    if (budget.remaining < 2) {
        return false;
    }
    var low = 0;
    var high = Math.min(value.length, budget.remaining - 2);
    while (low < high) {
        var mid = Math.ceil((low + high) / 2);
        if (JSON.stringify(value.slice(0, mid)).length <= budget.remaining) {
            low = mid;
        }
        else {
            high = mid - 1;
        }
    }
    return appendWithLimit(parts, JSON.stringify(value.slice(0, low)), budget);
};
var isJSONSerializablePrimitive = function (value) {
    return typeof value !== 'undefined' &&
        typeof value !== 'function' &&
        typeof value !== 'symbol' &&
        typeof value !== 'bigint';
};
var isNumberOrBoolean = function (value) {
    try {
        return (0, core_1.isNumber)(value) || (0, core_1.isBoolean)(value);
    }
    catch (_a) {
        return false;
    }
};
var stringifyValueWithLimit = function (value, parts, budget, seen, inArray) {
    if (inArray === void 0) { inArray = false; }
    if (!isJSONSerializablePrimitive(value)) {
        return inArray ? appendWithLimit(parts, 'null', budget) : true;
    }
    if ((0, core_1.isNull)(value) || isNumberOrBoolean(value)) {
        return appendWithLimit(parts, JSON.stringify(value), budget);
    }
    if (typeof value === 'string') {
        return stringifyStringWithLimit(value, parts, budget);
    }
    if (!(0, core_1.isObject)(value) && !(0, core_1.isArray)(value)) {
        return appendWithLimit(parts, JSON.stringify(value), budget);
    }
    if (seen.has(value)) {
        return stringifyStringWithLimit('[Circular]', parts, budget);
    }
    seen.add(value);
    try {
        var toJSON = value.toJSON;
        if ((0, core_1.isFunction)(toJSON)) {
            return stringifyValueWithLimit(toJSON.call(value), parts, budget, seen, inArray);
        }
    }
    catch (_a) {
        // If toJSON can't be read or throws, fall through to safe property enumeration.
    }
    try {
        var objectTag = Object.prototype.toString.call(value);
        if (objectTag === '[object String]') {
            return stringifyStringWithLimit(String(value.valueOf()), parts, budget);
        }
        if (objectTag === '[object Number]' || objectTag === '[object Boolean]') {
            return appendWithLimit(parts, JSON.stringify(value.valueOf()), budget);
        }
    }
    catch (_b) {
        // If Object.prototype.toString or valueOf throws, fall through to safe property enumeration.
    }
    if (value instanceof Error) {
        var errorObject = {};
        try {
            for (var key in value) {
                if (Object.prototype.hasOwnProperty.call(value, key)) {
                    errorObject[key] = value[key];
                }
            }
        }
        catch (_c) { }
        try {
            errorObject.name = value.name;
        }
        catch (_d) { }
        try {
            errorObject.message = value.message;
        }
        catch (_e) { }
        try {
            errorObject.stack = value.stack;
        }
        catch (_f) { }
        return stringifyValueWithLimit(errorObject, parts, budget, seen, inArray);
    }
    if ((0, core_1.isArray)(value)) {
        if (!appendWithLimit(parts, '[', budget)) {
            return false;
        }
        for (var i = 0; i < value.length; i++) {
            if (i > 0 && !appendWithLimit(parts, ',', budget)) {
                return false;
            }
            var item = void 0;
            try {
                item = value[i];
            }
            catch (_g) {
                item = undefined;
            }
            if (!stringifyValueWithLimit(item, parts, budget, seen, true)) {
                return false;
            }
        }
        return appendWithLimit(parts, ']', budget);
    }
    if (!appendWithLimit(parts, '{', budget)) {
        return false;
    }
    var isFirst = true;
    try {
        for (var key in value) {
            if (!Object.prototype.hasOwnProperty.call(value, key)) {
                continue;
            }
            if (budget.remaining <= 0) {
                budget.truncated = true;
                return false;
            }
            var propertyValue = void 0;
            try {
                propertyValue = value[key];
            }
            catch (_h) {
                continue;
            }
            if (!isJSONSerializablePrimitive(propertyValue)) {
                continue;
            }
            var propertyPrefix = "".concat(isFirst ? '' : ',').concat(JSON.stringify(key), ":");
            if (propertyPrefix.length >= budget.remaining) {
                budget.truncated = true;
                return false;
            }
            var partsBeforeProperty = parts.length;
            var remainingBeforeProperty = budget.remaining;
            var truncatedBeforeProperty = budget.truncated;
            if (!appendWithLimit(parts, propertyPrefix, budget)) {
                return false;
            }
            var partsBeforeValue = parts.length;
            var serialized = stringifyValueWithLimit(propertyValue, parts, budget, seen, false);
            var truncatedAfterValue = budget.truncated;
            if (parts.length === partsBeforeValue) {
                parts.length = partsBeforeProperty;
                budget.remaining = remainingBeforeProperty;
                budget.truncated = serialized ? truncatedBeforeProperty : truncatedAfterValue;
                if (!serialized) {
                    return false;
                }
                continue;
            }
            isFirst = false;
            if (!serialized) {
                return false;
            }
        }
    }
    catch (_j) {
        // we'll omit this object's properties considering we can't enumerate them
    }
    return appendWithLimit(parts, '}', budget);
};
var stringifyArgsSafely = function (args, sizeLimit) {
    var parts = [];
    var budget = { remaining: sizeLimit, truncated: false };
    for (var i = 0; i < args.length; i++) {
        if (i > 0 && !appendWithLimit(parts, ' ', budget)) {
            break;
        }
        if (!stringifyValueWithLimit(args[i], parts, budget, new WeakSet())) {
            break;
        }
    }
    return {
        body: parts.join('') + (budget.truncated ? '...' : ''),
        truncated: budget.truncated,
    };
};
/**
 * Flattens a nested object into a single level dot-notation object.
 * By default limit to 200kB or 50 keys.
 */
var flattenObject = function (obj, prefix, result, keys_limit, size_limit, seen) {
    if (prefix === void 0) { prefix = ''; }
    if (result === void 0) { result = {}; }
    if (keys_limit === void 0) { keys_limit = LOG_ATTRIBUTES_LIMIT; }
    if (size_limit === void 0) { size_limit = LOG_BODY_SIZE_LIMIT; }
    if (seen === void 0) { seen = new WeakSet(); }
    if (seen.has(obj)) {
        result[prefix || 'circular'] = '[Circular]';
        return result;
    }
    seen.add(obj);
    try {
        for (var key in obj) {
            try {
                if (!Object.prototype.hasOwnProperty.call(obj, key)) {
                    continue;
                }
                var value = obj[key];
                var newKey = prefix ? "".concat(prefix, ".").concat(key) : key;
                if ((0, core_1.isObject)(value)) {
                    flattenObject(value, newKey, result, keys_limit, size_limit, seen);
                }
                else {
                    keys_limit -= 1;
                    size_limit -= String(value).length;
                    size_limit -= newKey.length;
                    if (keys_limit <= 0 || size_limit <= 0) {
                        result['attributes_truncated'] = true;
                        return;
                    }
                    else {
                        result[newKey] = value;
                    }
                }
            }
            catch (_a) {
                continue;
            }
        }
    }
    catch (_b) {
        // we'll omit this object's properties considering we can't enumerate them
    }
    return result;
};
var SEVERITY_MAP = {
    log: 'INFO',
    warn: 'WARNING',
    error: 'ERROR',
    debug: 'DEBUG',
    info: 'INFO',
};
var initializeLogs = function (posthog) {
    var e_1, _a;
    setupOpenTelemetry(posthog);
    var logger = api_logs_1.logs.getLogger('console');
    var attributes = {};
    if (posthog.sessionManager) {
        var _b = posthog.sessionManager.checkAndGetSessionAndWindowId(true), sessionStartTimestamp = _b.sessionStartTimestamp, lastActivityTimestamp = _b.lastActivityTimestamp;
        attributes = {
            sessionStartTimestamp: sessionStartTimestamp.toString(),
            lastActivityTimestamp: lastActivityTimestamp.toString(),
        };
    }
    var _loop_1 = function (level) {
        var logWrapper = function (originalConsoleLog) {
            return function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                if (args.length === 0) {
                    return;
                }
                if (!posthog.is_capturing()) {
                    originalConsoleLog.apply(globals_1.assignableWindow.console, args);
                    return;
                }
                var _a = stringifyArgsSafely(args, LOG_BODY_SIZE_LIMIT), body = _a.body, truncated = _a.truncated;
                var logAttributes = __assign(__assign({}, attributes), (truncated ? { body_truncated: 'true' } : {}));
                logger.emit({
                    severityText: SEVERITY_MAP[level],
                    body: body,
                    attributes: __assign(__assign({ 'log.source': "console.".concat(level), distinct_id: posthog.get_distinct_id(), 'location.href': globals_1.assignableWindow.location.href }, logAttributes), ((0, core_1.isObject)(args[0]) ? flattenObject(args[0]) : {})),
                });
                originalConsoleLog.apply(globals_1.assignableWindow.console, args);
            };
        };
        var originalConsoleLog = globals_1.assignableWindow.console[level];
        globals_1.assignableWindow.console[level] = logWrapper(originalConsoleLog);
    };
    try {
        for (var _c = __values(['debug', 'log', 'warn', 'error', 'info']), _d = _c.next(); !_d.done; _d = _c.next()) {
            var level = _d.value;
            _loop_1(level);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
        }
        finally { if (e_1) throw e_1.error; }
    }
};
globals_1.assignableWindow.__PosthogExtensions__ = globals_1.assignableWindow.__PosthogExtensions__ || {};
globals_1.assignableWindow.__PosthogExtensions__.logs = { initializeLogs: initializeLogs };
